### PR TITLE
[OPS-324] Job for building downstream dependencies

### PR DESCRIPTION
**Pull request by glewis-ANet**
* Kick off the pipeline for all downstream dependencies that are specified, defaulting to using the 'dev' branch due to legacy projects.

### CODE REVIEW

File: src/commands/build-downstream.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    type: string
+    default: dev
+
+steps:
+  - run:
+      name: Starting downstream build
+      environment:
+        PROJECTS: <<parameters.project>>
+        BRANCH: <<parameters.branch>>
+      # The script implicitly expects the token to be in $CIRCLECI_API_TOKEN
```
Comment (glewis-ANet) [inline src/commands/build-downstream.yml:23]: CIRCLECI_API_TOKEN is currently set in the context, so packages that use contexts don't have to set it manually in their own environment

**Comment by glewis-ANet** [comment]
This is a slightly enhanced version of what was previously done in the individual anet-parent CircleCI configuration.  This version copes with more than one downstream dependency being specified as well as printing out the URLs for the build jobs it has started.

The biggest lack, IMO, is that the downstream jobs are not connected to the original PR in an easy to digest way.  Rather you have look through the CircleCI output to see them, or be looking at the CircleCI dashboard when those builds are kicked off.  It may be possible to hook them to the GitHub PR, but we need some research on how to do so.

File: src/scripts/create_circle_build.sh [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		--header 'Content-Type: application/json')
+	if [ -n "${BRANCH-}" ]; then
+		CURL_CMD+=(--data "{\"branch\":\"${BRANCH}\"}")
+	fi
+	CURL_CMD+=("https://circleci.com/api/v2/project/gh/AchievementNetwork/${project}/pipeline")
+	STATUS_CODE=$("${CURL_CMD[@]}")
+
+	case "${STATUS_CODE}" in
+		2*)
+			# If the build request was successful, look up the workflow that was created
```
Comment (glewis-ANet) [inline src/scripts/create_circle_build.sh:40]: It's pretty painful that the request doesn't spit this back and we have to make another call to get it.

File: src/commands/build-downstream.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      The branch of the project(s) to build.
+      The same branch will be used for all projects specified.
+    type: string
+    default: dev
+
+steps:
+  - run:
+      name: Starting downstream build
+      environment:
+        PROJECTS: <<parameters.project>>
```
Comment (pabbott-ANet) [inline src/commands/build-downstream.yml:21]: Interesting. Don't think I've seen double angle brackets for parameter substitution before

File: src/scripts/create_circle_build.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+
+set -o pipefail
+set -u
+
+TMP_FILES=()
+trap 'if [ ${#TMP_FILES[@]} -ne 0 ]; then rm "${TMP_FILES[@]}"; fi' EXIT
```
Comment (pabbott-ANet) [inline src/scripts/create_circle_build.sh:7]: What does trap do in this case? Makes it not abort if it fails?

**Review by pabbott-ANet — APPROVED** [review]
Don't see any issues

File: src/commands/build-downstream.yml (hunk 3 above)
Comment (glewis-ANet) [reply src/commands/build-downstream.yml:21]: This is CircleCI specific.  Parameters are substituted into commands using double angle brackets.

They don't really explain it well, but there are examples in https://circleci.com/docs/2.0/reusing-config/ if you're interested

File: src/scripts/create_circle_build.sh (hunk 4 above)
Comment (glewis-ANet) [reply src/scripts/create_circle_build.sh:7]: `trap` is used here to run something when the script exits, no matter where and how that exit happens (caveat, it won't be run if the script terminates due to being sent a SIGKILL).  Deleting temporary files is a pretty common use for it :).  https://www.linuxjournal.com/content/bash-trap-command has a reasonable explanation.

**Pull request merged**